### DISCUSSION TITLE
tsconfig's allowImportingTsExtensions and the transpilation result - is there a way to set imported .ts to .js?

### DISCUSSION THREAD

**Discussion by mkvlrn-cm42** [swc-project/swc#9016] [Q&A]
I'm not even sure if this is a good practice now or not, but I've just only found out about `allowImportingTsExtensions` in `tsconfig.json`, which allows me to import my esm modules using `.ts` instead of `.js`, which makes a bit more sense during development.

So when transpiling with swc the resulting js files keep the imports as `.ts`.

Is there a setting in the cli or `.swcrc` that also transforms the extension of the imports from `.ts` to `.js`?

**Comment by kdy1** [comment]
Search for `resolveFully` in the docs. It's jsc.module.resolveFully

**Reply by falkenhawk** [reply]
it doesn't seem to rewrite `.ts` to `.js` in dynamic imports still, despite enabling that option...

**Reply by alshdavid** [reply]
Likewise here, I am using the experimental built-in TS type stripping in Nodejs and am trying to use SWC to rewrite the file extensions from `.ts` to `.js` so I can publish my library.

Using this as the build command: 
```
npx swc . -d dist \
  --no-swcrc \
  -C jsc.target=esnext \
  -C module.resolveFully=true \
  -C module.type="nodenext"
```

Apparently `jsc.module.resolveFully` does not exist but [module.resolveFully](https://swc.rs/docs/configuration/modules#resolvefully) does

The extensions are not rewritten to `.js`

<img width="391" alt="image" src="https://github.com/user-attachments/assets/5544367a-8d6e-4afa-a9a6-ddd810ffd4c5">

**Reply by mkvlrn-cm42** [reply]
Oh, I never came back to this discussion after figuring it out! After digging around a lot, the only thing that worked for me was installing the plugin `@swc/plugin-transform-imports` and using it like this in `.swcrc`:

```json
{
  "jsc": {
    "experimental": {
      "plugins": [
        [
          "@swc/plugin-transform-imports",
          {
            "^(.*?)\\.ts(x)?$": {
              "skipDefaultConversion": true,
              "transform": "{{matches.[1]}}.js"
            }
          }
        ]
      ]
    }
  }
}
```

`module.resolveFully` didn't change anything.

**Reply by alshdavid** [reply]
@mkvlrn, you're a legend! Thank you 🙏

**Reply by falkenhawk** [reply]
Thanks @mkvlrn ! But it seems `@swc/plugin-transform-imports` does not handle dynamic imports (`await import('foo.ts')`) either... 😭

**Reply by mkvlrn-cm42** [reply]
@falkenhawk I have a similar issue when using Workers that need to have a file name used instead of importing something the "usual" way, and I settled with doing some trickery:

```ts
const extension = import.meta.url.split(".").pop();
const worker = new Worker(`my-worker.${extension}`);
```

Works beautifully in my case, maybe you can use that as a starting point?

**Reply by lukakostic** [reply]
> Search for `resolveFully` in the docs. It's jsc.module.resolveFully

`module.resolveFully = true`  does nothing, import extensions remain `.ts`.

Suggested answer is using `@swc/plugin-transform-imports` , why must we resort to plugins? Why not always transform .ts to .js if target is es ?

**Reply by kdy1** [reply]
You don't need a plugin anymore. https://github.com/swc-project/swc/pull/9784

**Reply by prolific** [reply]
@kdy1 Can you share how to do it without the plugin now?

I used this config:
```json
"module": {
    "type": "es6",
    "resolveFully": true,
    "outFileExtension": "js"
  },
  ```
  
  and I am trying to transform the following code where imports are using `.tsx` to `.js` but it seems to be not working
  ```ts
import { Route as rootRoute } from "./routes/__root.tsx";
import { Route as IdImport } from "./routes/$id.tsx";
import { Route as IndexImport } from "./routes/index.tsx";
```

Package Version:
```json
"@swc/cli": "^0.6.0",
"@swc/core": "^1.10.9",
```

`outFileExtension` also doesn't seem to be documented

**Reply by dionysiusmarquis** [reply]
I have the same problem "@swc/cli": "0.7.8" and "@swc/core": "1.12.11". I cannot find a way to change the imports from .ts to .js. Also tried "resolveFully": true and "outFileExtension": "js"

**Reply by Sylphide** [reply]
I am also interested in a solution without using the plugin.

**Reply by dionysiusmarquis** [reply]
After I tested different settings and couldn‘t find a way to get it working as described in the docs I created a bug report some days ago: https://github.com/swc-project/swc/issues/10848

**Reply by alshdavid** [reply]
Likewise, I have been using the plugin to achieve this and have not yet found a way to do so without it. 

`outFileExtension` changes the output file extension without rewriting the imports - I haven't tried `resolveFully` yet
